Tests/AdyenTests/Adyen Tests/Components/Gift Card/GiftCardComponentTests.swift

```diff
@@ -0,0 +1,414 @@
+
+        delegateMock.onDidFail = { error, component in
+            XCTFail("delegateMock.onDidFail shouldn't be reported back to merchant.")
+        }
+
+        let onCheckBalanceExpectation = expectation(description: "Expect partialPaymentDelegate.onCheckBalance to be called.")
+        let balance = Balance(availableAmount: .init(value: 200, currencyCode: "EUR"), transactionLimit: .init(value: 1000, currencyCode: "EUR"))
+        partialPaymentDelegate.onCheckBalance = {
+            $2(.success(balance))
+            onCheckBalanceExpectation.fulfill()
+        }
+
+        UIApplication.shared.keyWindow?.rootViewController = sut.viewController
+
+        wait(for: .seconds(1))
+
+        XCTAssertTrue(errorView!.isHidden)
+
+        populate(cardNumber: "60643650100000000000", pin: "73737")
+
+        payButtonItemViewButton?.sendActions(for: .touchUpInside)
+
+        wait(for: .seconds(1))
+
+        XCTAssertFalse(errorView!.isHidden)
+        XCTAssertEqual(sut.errorItem.message, "Gift cards are only valid in the currency they were issued in")
+
+        waitForExpectations(timeout: 10, handler: nil)
+    }
+
+    func testZeroBalance() throws {
+        sut.payment = .init(amount: .init(value: 10, currencyCode: "EUR"), countryCode: "US")
+
+        delegateMock.onDidFail = { error, component in
+            XCTFail("delegateMock.onDidFail shouldn't be reported back to merchant.")
+        }
+
+        let onCheckBalanceExpectation = expectation(description: "Expect partialPaymentDelegate.onCheckBalance to be called.")
+        let balance = Balance(availableAmount: .init(value: 0, currencyCode: "EUR"), transactionLimit: .init(value: 1000, currencyCode: "EUR"))
+        partialPaymentDelegate.onCheckBalance = {
+            $2(.success(balance))
+            onCheckBalanceExpectation.fulfill()
+        }
+
+        UIApplication.shared.keyWindow?.rootViewController = sut.viewController
+
+        wait(for: .seconds(1))
+
+        XCTAssertTrue(errorView!.isHidden)
+
+        populate(cardNumber: "60643650100000000000", pin: "73737")
+
+        payButtonItemViewButton?.sendActions(for: .touchUpInside)
+
+        wait(for: .seconds(1))
+
+        XCTAssertFalse(errorView!.isHidden)
+        XCTAssertEqual(sut.errorItem.message, "This gift card has zero balance")
+
+        waitForExpectations(timeout: 10, handler: nil)
+    }
+
+    func testMissingPaymentObject() throws {
+        sut.payment = nil // Missing Payment object
+
+        delegateMock.onDidFail = { error, component in
+            XCTFail("delegateMock.onDidFail shouldn't be reported back to merchant.")
+        }
+
+        let onCheckBalanceExpectation = expectation(description: "Expect partialPaymentDelegate.onCheckBalance to be called.")
+        let balance = Balance(availableAmount: .init(value: 200, currencyCode: "EUR"), transactionLimit: .init(value: 1000, currencyCode: "EUR"))
+        partialPaymentDelegate.onCheckBalance = {
+            $2(.success(balance))
```

Comment:
i think if there are three input parameters for a closure it makes sense to have names for them (at least the relevant ones). ` { _, _, completion in ... }`